core/src/Cabana_CommunicationPlan.hpp

```diff
@@ -26,6 +28,150 @@
+
+        // Do the duplicated fetch-add. If this is a valid element id
+        // increment the send count for this rank. Add the incremented count
+        // as the thread-local neighbor id. This is too big by one (because we
+        // use the prefix increment operator) but we want a non-zero value so
+        // we can later find which thread this element was located on because
+        // we are always guaranteed a non-zero value. We will subtract this
+        // value later.
+        if ( _element_export_ranks(i) >= 0 )
+            _neighbor_ids_dup(thread_id,i) =
+                ++_neighbor_counts_dup( thread_id, _element_export_ranks(i) );
+
+        // Release the thread id.
+        _unique_token.release( thread_id );
+    }
+
+    void reduce()
+    {
+        // Compute the send counts for each neighbor rank by reducing across
+        // the thread duplicates.
+        Kokkos::parallel_for(
+            Kokkos::RangePolicy<ExecutionSpace>(0,_neighbor_counts.extent(0)),
+            KOKKOS_LAMBDA( const size_type i ){
+                // Initialize with thread 0.
+                _neighbor_counts(i) = _neighbor_counts_dup(0,i);
+
+                // Add the rest of the thread results.
+                for ( size_type d = 1; d < _neighbor_counts_dup.extent(0); ++d )
+                    _neighbor_counts(i) += _neighbor_counts_dup(d,i);
+            });
+
+        // Compute the location of each export element in the send buffer of
+        // its destination rank.
+        Kokkos::parallel_for(
+            Kokkos::RangePolicy<ExecutionSpace>(
+                0,_element_export_ranks.extent(0)),
+            KOKKOS_LAMBDA( const size_type i ){
+                // Only operate on valid elements
+                if ( _element_export_ranks(i) >= 0 )
+                {
+                    // Start by determining the thread on which this element was
+                    // counted. We should only have a non-zero entry for the
+                    // neighbor id in the duplicate for the thread that
+                    // operated on it.
+                    int dup_thread = -1;
+                    for ( size_type d = 0; d < _neighbor_ids_dup.extent(0); ++d )
+                        if ( _neighbor_ids_dup(d,i) > 0 )
+                        {
+                            dup_thread = d;
+                            break;
+                        }
+
+                    // Next add the thread-local offset for this element. This
+                    // is where we subtract the 1 that we added artificially
+                    // when we were first counting.
+                    _neighbor_ids(i) = _neighbor_ids_dup(dup_thread,i) - 1;
+
+                    // Finally sum the thread counts prior to the one we
+                    // found.
+                    for ( int d = 0; d < dup_thread; ++d )
+                        _neighbor_ids(i) +=
+                            _neighbor_counts_dup(d,_element_export_ranks(i));
```

Comment:
Reduction over subview (if we want to use hierarchical parallelism here as well)

Reply:
refactored this segment of the code. I'll have to think of a way to use hierarchical parallelism with the new code

Reply:
was able to do hierarchical here as well.